File: server/routes/addresses.js
import { Router } from 'express'
import { query, withTransaction } from '../lib/db.js'
import { requireAuth } from '../lib/auth.js'

export const addressesRouter = Router()
addressesRouter.use(requireAuth) // all address routes require a logged-in user

// GET /api/addresses — the current user's saved delivery locations
addressesRouter.get('/', async (req, res) => {
  const { rows } = await query(
    'select * from addresses where user_id = $1 order by is_default desc, created_at',
    [req.user.id]
  )
  res.json({ addresses: rows })
})

// POST /api/addresses — add a delivery location
addressesRouter.post('/', async (req, res) => {
  const { label, city, street, house, notes, is_default } = req.body || {}
  if (!city || !street || !house) return res.status(400).json({ error: 'المدينة والشارع ورقم المنزل مطلوبة' })
  const address = await withTransaction(async (client) => {
    if (is_default) {
      await client.query('update addresses set is_default = false where user_id = $1', [req.user.id])
    }
    const { rows } = await client.query(
      `insert into addresses (user_id, label, city, street, house, notes, is_default)
       values ($1, $2, $3, $4, $5, $6, $7) returning *`,
      [req.user.id, label || null, city, street, house, notes || null, !!is_default]
    )
    return rows[0]
  })
  res.status(201).json({ address })
})

// PATCH /api/addresses/:id — edit one of the user's addresses
addressesRouter.patch('/:id', async (req, res) => {
  const allowed = ['label', 'city', 'street', 'house', 'notes', 'is_default']
  const fields = Object.keys(req.body || {}).filter((k) => allowed.includes(k))
  if (!fields.length) return res.status(400).json({ error: 'لا توجد حقول للتحديث' })
  const address = await withTransaction(async (client) => {
    if (req.body.is_default) {
      await client.query('update addresses set is_default = false where user_id = $1', [req.user.id])
    }
    const set = fields.map((f, i) => `${f} = $${i + 1}`).join(', ')
    const values = fields.map((f) => req.body[f])
    const { rows } = await client.query(
      `update addresses set ${set} where id = $${fields.length + 1} and user_id = $${fields.length + 2} returning *`,
      [...values, req.params.id, req.user.id]
    )
    return rows[0]
  })
  if (!address) return res.status(404).json({ error: 'العنوان غير موجود' })
  res.json({ address })
})

// DELETE /api/addresses/:id
addressesRouter.delete('/:id', async (req, res) => {
  const { rowCount } = await query('delete from addresses where id = $1 and user_id = $2', [req.params.id, req.user.id])
  if (!rowCount) return res.status(404).json({ error: 'العنوان غير موجود' })
  res.status(204).end()
})

File: backend/routers/addresses.py
from fastapi import APIRouter, Body, Depends, HTTPException, Request, Response

from db import pool, fetch_all, fetch_one
from security import current_user
from audit import log_action

router = APIRouter()


@router.get("")
def list_addresses(user=Depends(current_user)):
    rows = fetch_all("select * from addresses where user_id = %s order by is_default desc, created_at", [user["id"]])
    return {"addresses": rows}


@router.post("")
def add_address(request: Request, response: Response, user=Depends(current_user), payload: dict = Body(default={})):
    city, street, house = payload.get("city"), payload.get("street"), payload.get("house")
    if not city or not street or not house:
        raise HTTPException(400, "City, street and house number are required")
    is_default = bool(payload.get("is_default"))
    with pool.connection() as conn, conn.transaction(), conn.cursor() as cur:
        if is_default:
            cur.execute("update addresses set is_default = false where user_id = %s", [user["id"]])
        cur.execute(
            """insert into addresses (user_id, label, city, street, house, notes, is_default)
               values (%s, %s, %s, %s, %s, %s, %s) returning *""",
            [user["id"], payload.get("label"), city, street, house, payload.get("notes"), is_default],
        )
        address = cur.fetchone()
    log_action(user_id=user["id"], action="address_added", detail={"city": city}, request=request)
    response.status_code = 201
    return {"address": address}


@router.patch("/{aid}")
def update_address(aid: str, user=Depends(current_user), payload: dict = Body(default={})):
    allowed = ["label", "city", "street", "house", "notes", "is_default"]
    fields = [k for k in (payload or {}) if k in allowed]
    if not fields:
        raise HTTPException(400, "No fields to update")
    with pool.connection() as conn, conn.transaction(), conn.cursor() as cur:
        if payload.get("is_default"):
            cur.execute("update addresses set is_default = false where user_id = %s", [user["id"]])
        set_clause = ", ".join(f"{f} = %s" for f in fields)
        values = [payload[f] for f in fields] + [aid, user["id"]]
        cur.execute(f"update addresses set {set_clause} where id = %s and user_id = %s returning *", values)
        address = cur.fetchone()
    if not address:
        raise HTTPException(404, "Address not found")
    return {"address": address}


@router.delete("/{aid}")
def delete_address(aid: str, response: Response, request: Request, user=Depends(current_user)):
    row = fetch_one("delete from addresses where id = %s and user_id = %s returning id", [aid, user["id"]])
    if not row:
        raise HTTPException(404, "Address not found")
    log_action(user_id=user["id"], action="address_removed", request=request)
    response.status_code = 204
    return None
